File: android/src/test/java/com/facebook/spectrum/SpectrumExceptionTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.spectrum;

import static org.fest.assertions.api.Assertions.assertThat;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class SpectrumExceptionTest {

  public static final String EXCEPTION_MESSAGE = "Flux compensator implosion";
  public static final SpectrumException EXCEPTION_CAUSE = new SpectrumException("time drift");

  @Test
  public void testSpectrumException_whenGivenMessage_thenContainsMessage() {
    final SpectrumException spectrumException = new SpectrumException(EXCEPTION_MESSAGE);
    assertThat(spectrumException.getMessage()).isEqualTo(EXCEPTION_MESSAGE);
  }

  @Test
  public void testSpectrumException_whenGivenCause_thenContainsCause() {
    final SpectrumException spectrumException = new SpectrumException(EXCEPTION_CAUSE);
    assertThat(spectrumException.getCause()).isEqualTo(EXCEPTION_CAUSE);
  }

  @Test
  public void testSpectrumException_whenGivenMessageAndCause_thenContainsMessageAndCause() {
    final SpectrumException spectrumException =
        new SpectrumException(EXCEPTION_MESSAGE, EXCEPTION_CAUSE);
    assertThat(spectrumException.getMessage()).isEqualTo(EXCEPTION_MESSAGE);
    assertThat(spectrumException.getCause()).isEqualTo(EXCEPTION_CAUSE);
  }
}

File: cpp/test/unit/SpectrumExceptionTest.cpp
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

#include <spectrum/SpectrumException.h>

#include <gtest/gtest.h>

#include <cstring>

namespace facebook {
namespace spectrum {
namespace test {

TEST(SpectrumException, whenBuilding_nothingObservableHappens) {
  const auto name = "name";
  const auto message = std::string{"message"};
  const auto function = "function";
  const auto line = 42;

  const auto exception = SpectrumException{name, message, function, line};

  const auto expectedLocation = "function:42";
  const auto expectedDescription = "name#message-function:42";

  ASSERT_EQ(name, exception.name);
  ASSERT_EQ(message, exception.message);
  ASSERT_EQ(expectedLocation, exception.location);
  ASSERT_EQ(expectedDescription, exception.description);
  ASSERT_EQ(0, std::strcmp(expectedDescription, exception.what()));
}

} // namespace test
} // namespace spectrum
} // namespace facebook
